javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1216 @@
+      if (lo < hi) {
+        val mid = lo + (hi - lo) / 2
+        val cmp = compareToKey(get(mid))
+        if (cmp == 0) mid
+        else if (cmp > 0) binarySearch(lo, mid, get)
+        else binarySearch(mid + 1, hi, get)
+      } else {
+        notFound(lo)
+      }
+    }
+
+    list match {
+      case _: RandomAccess =>
+        binarySearch(0, list.size, list(_))
+
+      case _ =>
+        def getFrom(iter: ListIterator[E])(index: Int): E = {
+          val shift = index - iter.nextIndex
+          if (shift > 0)
+            (0 until shift).foreach(_ => iter.next())
+          else
+            (0 until -shift).foreach(_ => iter.previous())
+          iter.next()
+        }
+        binarySearch(0, list.size, getFrom(list.listIterator))
+    }
+  }
+
+  def reverse(list: List[_]): Unit =
+    reverseImpl(list)
+
+  @inline
+  def reverseImpl[T](list: List[T]): Unit = {
+    val size = list.size
+    list match {
+      case list: RandomAccess =>
+        for (i <- 0 until size / 2) {
+          val tmp = list(i)
+          list(i) = list(size - i - 1)
+          list(size - i - 1) = tmp
+        }
+
+      case _ =>
+        val it1 = list.listIterator()
+        val it2 = list.listIterator(size)
+        for (i <- 0 until size / 2) {
+          val tmp = it1.next()
+          it1.set(it2.previous())
+          it2.set(tmp)
+        }
+    }
+  }
+
+  def shuffle(list: List[_]): Unit =
+    shuffle(list, new Random)
+
+  def shuffle(list: List[_], rnd: Random): Unit =
+    shuffleImpl(list, rnd)
+
+  @inline
+  def shuffleImpl[T](list: List[T], rnd: Random): Unit = {
+    val scalaRnd = scala.util.Random.javaRandomToRandom(rnd)
+    val shuffledList = scalaRnd.shuffle(list.toSeq)
+    list match {
+      case list: RandomAccess => copyImpl(shuffledList.iterator, list)
+      case _                  => copyImpl(shuffledList.iterator, list.listIterator)
+    }
+  }
+
+  def swap(list: List[_], i: Int, j: Int): Unit =
+    swapImpl(list, i, j)
+
+  @inline
+  private def swapImpl[E](list: List[E], i: Int, j: Int): Unit = {
+    list match {
+      case list: RandomAccess =>
+        val tmp = list(i)
+        list(i) = list(j)
+        list(j) = tmp
+
+      case _ =>
+        val it1 = list.listIterator(i)
+        val it2 = list.listIterator(j)
+        if (!it1.hasNext || !it2.hasNext)
+          throw new IndexOutOfBoundsException
+        val tmp = it1.next()
+        it1.set(it2.next())
+        it2.set(tmp)
+    }
+  }
+
+  def fill[T](list: List[_ >: T], obj: T): Unit = {
+    list match {
+      case list: RandomAccess =>
+        (0 until list.size).foreach(list(_) = obj)
+
+      case _ =>
+        val iter = list.listIterator
+        while (iter.hasNext) {
+          iter.next()
+          iter.set(obj)
+        }
+    }
+  }
+
+  def copy[T](dest: List[_ >: T], src: List[_ <: T]): Unit = {
+    (dest, src) match {
+      case (dest: RandomAccess, src: RandomAccess) => copyImpl(src, dest)
```

Comment:
Wow. I'm impressed that scalac considers that `dest` has type `List[_ <: T] with RandomAccess` ^^